### PR TITLE
Support multi-objective optimization in `WeightsAndBiasesCallback`

### PR DESCRIPTION
**Pull request by xadrianzetx**
This PR introduces support for tracking multi-objective optimization in Weights & Biases UI by extending functionality of `WeightsAndBiasesCallback`

Now, arbitrary number of trial values can be tracked with distinctive names, and optimization directions.

## Motivation
Follow up to [#2781 review](https://github.com/optuna/optuna/pull/2781#pullrequestreview-701578675)

## Description of the changes

*  Modify `WeightsAndBiasesCallback` to support multi-objective optimization
*  Modify and parametrize some of existing tests to facilitate the change
*  Add new test cases

### Objective function
Using modified version of objective function from #2553 with two objectives.

```python
# [...]

def objective(self, trial):
    # Parameters.
    trial_alpha = trial.suggest_uniform("alpha", low=-10, high=10)
    trial_beta = trial.suggest_uniform("beta", low=-10, high=10)

    # Prediction and loss.
    y_hat = self.X1 * trial_alpha + self.X2 * trial_beta
    mse = ((self.y - y_hat) ** 2).mean()
    mae = (np.abs(self.y - y_hat)).mean()

    return mse,  mae

# [...]
```
### Optimization

```python
black_box = Objective(seed=4444)
sampler = optuna.samplers.TPESampler(seed=4444)

wandb_kwargs = {"project": "optuna"}

wandbc = WeightsAndBiasesCallback(
    metric_name=["mse", "mae"],
    wandb_kwargs=wandb_kwargs
)

study = optuna.create_study(directions=["minimize", "minimize"], sampler=sampler)
study.optimize(black_box, n_trials=100, callbacks=[wandbc])
```
### W&B UI results

image

image

### CODE REVIEW

File: tests/integration_tests/test_wandb.py [hunk 1]
```diff
@@ -59,15 +87,49 @@ def test_log_api_call_count(wandb: mock.Mock) -> None:
...
+    wandbc = WeightsAndBiasesCallback(metric_name=metrics)
+    study = optuna.create_study(directions=["minimize", "maximize"])
+    study.optimize(_multiobjective_func, n_trials=1, callbacks=[wandbc])
+
+    kall = wandb.log.call_args
+    assert list(kall[0][0].keys()) == expected
+    assert kall[1] == {"step": 0}
+
+
+@pytest.mark.parametrize("metrics", [(["foo"]), (["foo", "bar", "baz"])])
```
Comment (nzw0301) [inline tests/integration_tests/test_wandb.py:127]:
```suggestion
@pytest.mark.parametrize("metrics", [["foo"], ["foo", "bar", "baz"]])
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Thank you for sending the follow-up PR! Let me leave a minor suggestion.

File: tests/integration_tests/test_wandb.py (hunk 1 above)
Comment (xadrianzetx) [reply tests/integration_tests/test_wandb.py]: Thanks! Addressed in b3bc6f718ab0c879ec86d95cf40ef368f2b342a0

**Review by nzw0301 — APPROVED** [review]
LGTM! Thanks!

File: optuna/integration/wandb.py [hunk 2]
```diff
@@ -84,11 +89,32 @@ def __init__(
...
+                    )
+                )
+
+            else:
+                names = self._metric_name
+
+        if isinstance(self._metric_name, str):
+            if len(trial.values) > 1:
+                # broadcast default name for multi-objective optimization
+                names = ["{}_{}".format(self._metric_name, i) for i in range(len(trial.values))]
```
Comment (HideakiImamura) [inline optuna/integration/wandb.py:108]: I think we should write this "broadcast" logic in the document.

Comment (HideakiImamura) [reply optuna/integration/wandb.py:108]: In addition, how about testing this logic?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the follow-up PR! I have several comments. Could you take a look?

File: optuna/integration/wandb.py (hunk 2 above)
Comment (xadrianzetx) [reply optuna/integration/wandb.py:108]: Hi @HideakiImamura, thanks for the feedback! We are testing this logic in 72f7071b1a51d6ae082071bf2165a0e9a60a1b51. Do you think there should be more test cases covering this?

Comment (xadrianzetx) [reply optuna/integration/wandb.py:112]: Pushed documentation update on broadcasting in 5cb2d80f229bd305437bfda39984e3038a859f3c.

**Comment by xadrianzetx** [comment]
Seems like GitHub Actions is having some problems.

image

**Comment by nzw0301** [comment]
Thank you for letting us know about the CI error. I re-ran the CI jobs. I hope it works now 🤞🏻

**Comment by xadrianzetx** [comment]
Seems like those jobs are still stuck. @nzw0301, i can try to merge master to this branch and trigger them again if it's ok with you.

**Comment by nzw0301** [comment]
@xadrianzetx Sure! Thank you for your suggestion.

File: optuna/integration/wandb.py (hunk 2 above)
Comment (HideakiImamura) [reply optuna/integration/wandb.py:112]: It certainly looks like it has been tested enough. Thanks!

**Review by HideakiImamura — APPROVED** [review]
LGTM!

**Pull request merged**